src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,133 @@
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// Defines Extension methods which are meant to be re-used across WinHttp & UnixHttp Handlers
+    /// </summary>
+    internal static class TLSCertificateExtensions
+    {
+        private const string ClientCertificateOid = "1.3.6.1.5.5.7.3.2";
+
+        /// <summary>
+        ///   returns true if the X509 Certificate can be used  as SSL Client Certificate.
+        /// </summary>
+        private static bool IsClientCertificate(X509Certificate2 cert)
+        {
+            Debug.Assert(cert != null, "certificate cannot be null");
+
+            bool hasClientCertExt = false;
+            bool forDigitalSignature = false;
+            bool noExtendedRestrictions = true;
+            bool noUsageRestrictions = true;
+            foreach(X509Extension extension in cert.Extensions)
+            {
+                // check if the extension is an enhanced usage ext.
+                // But do this only if needed. No point going over it, if we already have established that our cert has the
+                // required extension.
+                if (!hasClientCertExt)
+                {
+                    X509EnhancedKeyUsageExtension enhancedUsageExt = extension as X509EnhancedKeyUsageExtension;
+                    if (enhancedUsageExt != null)
+                    {
+                        noExtendedRestrictions = false;
+                        foreach(Oid oid in enhancedUsageExt.EnhancedKeyUsages)
+                        {
+                            if (string.Equals(ClientCertificateOid, oid.Value))
+                            {
+                                hasClientCertExt = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!forDigitalSignature)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        noUsageRestrictions = false;
+                        forDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if(hasClientCertExt && forDigitalSignature)
+                {
+                    return true;
+                }
+            }
+            Debug.Assert(!hasClientCertExt || !forDigitalSignature, "if cert can be used as digital signature && it has ClientCertificateOid extension, it is a client cert");
+
+            return noExtendedRestrictions && noUsageRestrictions;
```

Comment:
This isn't quite correct.  You're assuming here that either both or neither of the extensions are present, but it could be one without the other (most likely KeyUsages without ExtendedKeyUsages).

Using your current variable names/purposes, you should have something like

```C#
hasClientCertExt |= noExtendedRestrictions;
forDigitalSignature |= noUsageRestrictions;

return hasClientCertExt && forDigitalSignature;
```

I'd change the method to
* use `foundEku`, `foundKeyUsages`, `isClientAuth`, and `isDigitalSignature`
    * The `found` booleans starting off as false, the `is` as true.
    * When setting `foundEku` to true, set `isClientAuth` to false, then set it back to true if you find the match.
    * `isDigitalSignature` is easier, since it's correctly assignable in a single operation.
* change the early return to `if (foundEku && foundKeyUsages) { break; }`